lib/Dialect/Calyx/Transforms/CompileControl.cpp

```diff
@@ -0,0 +1,190 @@
+         "The SeqOp should only contain EnableOps in this pass.");
+  // This should be the number of enable statements + 1 since this is the
+  // maximum value the FSM register will reach.
+  size_t fsmBitWidth = getNecessaryBitWidth(seqOps.size() + 1);
+
+  OpBuilder builder(component->getRegion(0));
+  auto fsmRegister = createRegister(builder, component, fsmBitWidth, "fsm");
+  auto fsmIn = fsmRegister.getResult(0);
+  auto fsmWriteEn = fsmRegister.getResult(1);
+  auto fsmOut = fsmRegister.getResult(4);
+
+  builder.setInsertionPointToStart(wiresBody);
+  auto oneConstant = createConstant(builder, wires->getLoc(), 1, 1);
+
+  // Create the new compilation group to replace this SeqOp.
+  builder.setInsertionPointToEnd(wiresBody);
+  auto seqGroup =
+      builder.create<GroupOp>(wires->getLoc(), builder.getStringAttr("seq"));
+  Block *seqGroupBody = new Block();
+  seqGroup->getRegion(0).push_back(seqGroupBody);
+
+  // Guarantees a unique SymbolName for the group.
+  SymbolTable symTable(wires);
+  symTable.insert(seqGroup);
+
+  size_t fsmIndex = 0;
+  SmallVector<Attribute, 8> groupNames;
+  Value fsmNextState;
+  seq.walk([&](EnableOp enable) {
+    StringRef groupName = enable.groupName();
+    groupNames.push_back(SymbolRefAttr::get(builder.getContext(), groupName));
+    auto groupOp = cast<GroupOp>(symTable.lookup(groupName));
+
+    builder.setInsertionPoint(groupOp);
+    auto fsmCurrentState =
+        createConstant(builder, wires->getLoc(), fsmBitWidth, fsmIndex);
+
+    // TODO(Calyx): Eventually, we should canonicalize the GroupDoneOp's guard
+    // and source.
+    auto guard = groupOp.getDoneOp().guard();
+    auto source = groupOp.getDoneOp().src();
+    auto doneOpValue =
+        !guard ? source
+               : builder.create<comb::AndOp>(wires->getLoc(), guard, source);
+
+    // Build the Guard for the `go` signal of the current group being walked.
+    // The group should begin when:
+    // (1) the current step in the fsm is reached, and
+    // (2) the done signal of this group is not high.
+    auto eqCmp = builder.create<comb::ICmpOp>(
+        wires->getLoc(), comb::ICmpPredicate::eq, fsmOut, fsmCurrentState);
+    auto notDone =
+        builder.create<comb::XorOp>(wires->getLoc(), doneOpValue, oneConstant);
+    auto groupGoGuard =
+        builder.create<comb::AndOp>(wires->getLoc(), eqCmp, notDone);
+
+    // Guard for the `in` and `write_en` signal of the fsm register. These are
+    // driven when the group has completed.
+    builder.setInsertionPoint(seqGroup);
+    auto groupDoneGuard =
+        builder.create<comb::AndOp>(wires->getLoc(), eqCmp, doneOpValue);
+
+    // Directly update the GroupGoOp of the current group being walked.
+    auto goOp = groupOp.getGoOp();
+    assert(goOp && "The Go Insertion pass should be run before this.");
+    goOp->setOperand(0, oneConstant);
+    goOp->insertOperands(1, {groupGoGuard});
```

Comment:
Just curious, is there any reason not to just do `setOperand` here?

Reply:
Looking at the API again, 
```goOp->setOperands({oneConstant, groupGoGuard});``` 
is the cleanest approach to this IMO.